include/stl2/view/subrange.hpp

```diff
@@ -0,0 +1,237 @@
+			: subrange{__stl2::begin(r), __stl2::end(r)} {}
+
+			template <Constructible<const I&> X, Constructible<const S&> Y>
+			requires ConvertibleTo<const I&, X> && ConvertibleTo<const S&, Y>
+			constexpr operator pair<X, Y>() const {
+				return pair<X, Y>{this->begin(), this->end()};
+			}
+
+			constexpr bool empty() const {
+				return this->begin() == this->end();
+			}
+			constexpr explicit operator bool() const {
+				return !empty();
+			}
+			constexpr bool operator!() const {
+				return empty();
+			}
+			constexpr auto size() const requires Sized {
+				return this->end() - this->begin();
+			}
+		};
+
+		template <Iterator I, Sentinel<I> S>
+		requires !SizedSentinel<S, I>
+		struct subrange<I, S, true>
+		: view_interface<subrange<I, S, true>>
+		{
+		private:
+			subrange<I, S> rng_ {};
+			difference_type_t<I> size_ = 0;
+		public:
+			using iterator = I;
+			using sentinel = S;
+
+			subrange() = default;
+
+			constexpr subrange(I i, S s, difference_type_t<I> n)
+			: rng_{i, s}, size_{n} {}
+
+			template <ConvertibleTo<I> X, ConvertibleTo<S> Y>
+			constexpr subrange(subrange<X, Y, true> r)
+			: subrange{r.begin(), r.end(), r.size()} {}
+
+			template <ConvertibleTo<I> X, ConvertibleTo<S> Y, bool B>
+			constexpr subrange(subrange<X, Y, B> r, difference_type_t<I> n)
+			: subrange{r.begin(), r.end(), n} {
+				if constexpr (B) {
+					STL2_EXPECT(n == r.size());
+				}
+			}
+
+			template <ConvertibleTo<I> X, ConvertibleTo<S> Y>
+			constexpr subrange(pair<X, Y> r, difference_type_t<I> n)
+			: subrange{r.first, r.second, n} {}
+
+			template <SizedRange R>
+			requires Constructible<subrange<I, S>, R&>
+			constexpr subrange(R& r)
+			: subrange{__stl2::begin(r), __stl2::end(r), __stl2::size(r)} {}
+
+			template <Constructible<const I&> X, Constructible<const S&> Y>
+			requires ConvertibleTo<const I&, X> && ConvertibleTo<const S&, Y>
+			constexpr operator pair<X, Y>() const {
+				return pair<X, Y>{rng_.begin(), rng_.end()};
+			}
+
+			template <Constructible<const I&> X, Constructible<const S&> Y, bool B>
+			requires ConvertibleTo<const I&, X> && ConvertibleTo<const S&, Y>
+			constexpr operator subrange<X, Y, B>() const {
+				if constexpr (B) {
+					return {rng_, size_};
+				} else {
+					return rng_;
+				}
+			}
+
+			constexpr operator subrange<I, S> const &() const & noexcept {
+				return rng_;
+			}
+
+			constexpr I begin() const {
+				return rng_.begin();
+			}
+			constexpr S end() const {
+				return rng_.end();
+			}
+			constexpr bool empty() const {
+				return begin() == end();
+			}
+			constexpr explicit operator bool() const {
+				return !empty();
+			}
+			constexpr bool operator!() const {
+				return empty();
+			}
+			constexpr difference_type_t<I> size() const noexcept {
+				return size_;
+			}
+		};
+
+		template <Iterator I, Sentinel<I> S>
+		using sized_subrange = subrange<I, S, true>;
+
+		template <Iterator I, Sentinel<I> S>
+		subrange(I, S) -> subrange<I, S>;
```

Comment:
This should be redundant with the implicit guides.

Reply:
So because the `Sized` parameter is defaulted, it doesn't need to be deducible?

Reply:
Yes.